File: src/gestorAplicacion/gestion/Beca.java
package gestorAplicacion.gestion;

import java.io.Serializable;
import java.util.*;
import gestorAplicacion.gestion.*;
import gestorAplicacion.personas.*;

public class Beca implements Serializable {

	/**
	 * 
	 */
	private static final long serialVersionUID = 1L;
	private final float criterio;
	private int cupos;
	private ArrayList<Persona> beneficiarios;
	private HashMap<Facultad, Integer> cuposPorFacultad;
	public static ArrayList<Beca> listaBecas = new ArrayList<Beca>();
	private boolean isProfesoral;
	private String nombre;

	public Beca(String nombre, float criterio,
			HashMap<Facultad, Integer> cuposPorFacultad, boolean isProfesoral) {
		super();
		this.nombre = nombre;
		this.criterio = criterio;
		this.cuposPorFacultad = cuposPorFacultad;
		this.isProfesoral = isProfesoral;
		this.beneficiarios = new ArrayList<Persona>();
		this.setBeneficiarios();
		Beca.listaBecas.add(this);

	}

	private boolean Elegibilidad(Persona beneficiario) {
		float promedio = beneficiario.calcularPromedio();
		boolean result = (promedio >= criterio);
		return result;
	}

	public void setBeneficiarios() {

		HashMap<Facultad, Integer> cuposCopy = new HashMap<>(cuposPorFacultad);
		Collections.sort(Estudiante.getListaEstudiantes());
		Collections.sort(Profesor.getListaProfesores());

		for (Facultad facultad : cuposCopy.keySet()) {

			if (isProfesoral) {
				for (Profesor profesor : facultad.getProfesoresVinculados()) {
					boolean result = Elegibilidad(profesor);
					boolean cuposDisponibles = cuposCopy.get(facultad) > 0;
					if (result && cuposDisponibles) {
						beneficiarios.add(profesor);
						cuposCopy.put(facultad, cuposCopy.get(facultad) - 1);
					} else {
					}
				}

			} else {
				for (Estudiante estudiante : facultad.getEstudiantes()) {
					boolean result = Elegibilidad(estudiante);
					boolean cuposDisponibles = cuposCopy.get(facultad) > 0;
					if (result && cuposDisponibles) {
						beneficiarios.add(estudiante);
						cuposCopy.put(facultad, cuposCopy.get(facultad) - 1);
					} else {
					}
				}
			}
		}
	}

	public int cuposSobrantes() {
		int total = 0;
		for (Facultad f : cuposPorFacultad.keySet()) {
			total += cuposPorFacultad.get(f);
		}
		return total - beneficiarios.size();
	}

	public ArrayList<Persona> getBeneficiarios() {
		return beneficiarios;
	}

	public float promedioBeneficiarios() {
		float total = 0;
		for (Persona p : beneficiarios) {
			total += p.calcularPromedio();
		}
		return total / beneficiarios.size();
	}

	public float getCriterio() {
		return criterio;
	}

	public int getCupos() {
		return cupos;
	}

	public void setCupos(int cupos) {
		this.cupos = cupos;
	}

	public static ArrayList<Beca> getListaBecas() {
		return listaBecas;
	}

	public static void setListaBecas(ArrayList<Beca> listaBecas) {
		Beca.listaBecas = listaBecas;
	}

	public HashMap<Facultad, Integer> getCuposPorFacultad() {
		return cuposPorFacultad;
	}

	public void setCuposPorFacultad(HashMap<Facultad, Integer> cuposPorFacultad) {
		this.cuposPorFacultad = cuposPorFacultad;
	}

	public boolean isProfesoral() {
		return isProfesoral;
	}

	public void setProfesoral(boolean isProfesoral) {
		this.isProfesoral = isProfesoral;
	}

	public String getNombre() {
		return nombre;
	}

	public void setNombre(String nombre) {
		this.nombre = nombre;
	}

}

File: srcpy/gestorAplicacion/gestion/beca.py
class Beca:
    listaBecas = []

    def __init__(self, nombre, criterio, cuposPorFacultad, isProfesoral):
        self.nombre = nombre
        self.criterio = criterio
        self.cuposPorFacultad = cuposPorFacultad
        self.isprofesoral = isProfesoral
        self.setBeneficiarios()
        Beca.listaBecas.append(self)

    def Elegibilidad(self, beneficiario):
        from gestorAplicacion.personas.estudiante import Estudiante
        from gestorAplicacion.personas.profesor import Profesor

        promedio = beneficiario.calcularPromedio()
        result = promedio >= self.criterio
        return result

    def setBeneficiarios(self):
        from gestorAplicacion.personas.estudiante import Estudiante
        from gestorAplicacion.personas.profesor import Profesor
        from gestorAplicacion.gestion.facultad import Facultad

        Estudiante.getListaEstudiantes().sort()
        Profesor.getListaProfesores().sort()

        cuposCopy = self.cuposPorFacultad.copy()
        self.beneficiarios = []
        for facultad in cuposCopy.keys():
            if self.isProfesoral():
                for profesor in facultad.getProfesoresVinculados():
                    result = self.Elegibilidad(profesor)
                    cuposDisponibles = cuposCopy[facultad] > 0
                    if result and cuposDisponibles:
                        self.beneficiarios.append(profesor)
                        cuposCopy[facultad] = cuposCopy[facultad] - 1

            else:
                for estudiante in facultad.getEstudiantes():
                    result = self.Elegibilidad(estudiante)
                    cuposDisponibles = cuposCopy[facultad] > 0
                    if result and cuposDisponibles:
                        self.beneficiarios.append(estudiante)
                        cuposCopy[facultad] = cuposCopy[facultad] - 1

    def cuposSobrantes(self):
        total = 0
        for facultad in self.cuposPorFacultad.keys():
            total += self.cuposPorFacultad[facultad]
        return total - len(self.beneficiarios)

    def getBeneficiarios(self):
        return self.beneficiarios

    def promedioBeneficiarios(self):
        from gestorAplicacion.personas.estudiante import Estudiante
        from gestorAplicacion.personas.profesor import Profesor

        total = 0
        for persona in self.beneficiarios:
            total += persona.calcularPromedio()
        return total / len(self.beneficiarios)

    def getCriterio(self):
        return self.criterio

    def setCriterio(self, criterio):
        self.criterio = criterio

    def getCupos(self):
        return self.cupos

    def setCupos(self, cupos):
        self.cupos = cupos

    @classmethod
    def getListabecas(cls):
        return cls.listaBecas

    @classmethod
    def setListaBecas(cls, lista):
        cls.listaBecas = lista

    def getCuposPorFacultad(self):
        return self.cuposPorFacultad

    def setCuposPorFacultad(self, cupos):
        self.cuposPorFacultad = cupos

    def setProfesoral(self, isProfesoral):
        self.isprofesoral = isProfesoral

    def isProfesoral(self):
        return self.isprofesoral

    def getNombre(self):
        return self.nombre

    def setNombre(self, nombre):
        self.nombre = nombre
